### PR TITLE
Social First Signup: Better separation between User and UserSocial

### PR DESCRIPTION
**Pull request by escapemanuele**
## What

With https://github.com/Automattic/wp-calypso/pull/82481  we shipped the work done on pdDR7T-Z4-p2 .
We went from the usual signup step:
image

To the new Social First Signup:
image

## The problem

Signup is a central and core part of the whole WordPress.com existence, and we shipped a change that required more time to be built, tested, and thought than we expected.

Following excellent suggestions on how to improve this, I worked on a better structure for this change.

## Solution

While the previous implementation just changed the User step and the SignupForm component, this creates a new step, `user-social`, that can or not be added to a flow. 
This new step calls the old User step but tells it to be `social-first`.

Moreover, we have a new flag, `signup/social-first`, that can be used to disable the new step globally if needed.

## Testing
1. Click the Live link
2. You can check, while in **Incognito**,  the flows that are in [this array here](https://github.com/Automattic/wp-calypso/blob/update/componentize-signup-social-first/client/signup/config/flows-pure.js). The ones that use `userSocialStep` have the new signup step.
3. Check that flows like `onboarding-pm` are not affected by the new signup. (Visit `/start/onboarding-pm` )
4. Check flows from Stepper, like `/setup/newsletter`, `/setup/link-in-bio` and `/setup/domain-transfer`.

2. While in incognito, go through these flows. They should have the standard signup:

- Gravatar `/log-in?client_id=1854&redirect_to=https%3A%2F%2Fpublic-api.wordpress.com%2Foauth2%2Fauthorize%3Fclient_id%3D1854%26response_type%3Dcode%26blog_id%3D0%26state%3D500e79a3faed390d3ef5f2778ef39363706cc0609c02c3c25397c7499f3fa51c%26redirect_uri%3Dhttps%253A%252F%252Fen.gravatar.com%252Fconnect%252F%253Faction%253Drequest_access_token`
image
- P2: `/log-in?redirect_to=https%3A%2F%2Fwordpress.com%2Fstart%2Fp2%2Fp2-confirm-email%3Fref%3Dp2-lp&from=p2&signup_url=%2Fstart%2Fp2%2Fuser%3Fref%3Dp2-lp` 
image
-Crowdsignal: `/log-in?client_id=978&redirect_to=https%3A%2F%2Fpublic-api.wordpress.com%2Foauth2%2Fauthorize%3Fclient_id%3D978%26response_type%3Dcode%26blog_id%3D0%26state%3D51d5e1b0d7358dee3de65c45f4b4dc37a3d6e37dab8d6047545f636c2659d104%26redirect_uri%3Dhttps%253A%252F%252Fapp.crowdsignal.com%252Fconnect%253Fsource%253Dlogin%2526action%253Drequest_access_token%26wpcom_connect%3D1`
image
- Akismet: `/start/wpcc/oauth2-user?oauth2_client_id=973&oauth2_redirect=https%3A%2F%2Fpublic-api.wordpress.com%2Foauth2%2Fauthorize%3Fclient_id%3D973%26response_type%3Dcode%26blog_id%3D0%26state%3D4a9b8708620f66f5e004f26f53a9678dab2d731a4a03f3bbb962415a3b0501cf%26redirect_uri%3Dhttps%253A%252F%252Fakismet.com%252Faccount%252F%253Faction%253Drequest_access_token%26implicit%3Dtrue%26new-user%3D0&user_email=`
image
- Jetpack connect: `/log-in/jetpack?site=https%3A%2F%2Ffalse-shrike.jurassic.ninja&redirect_to=http%3A%2F%2Fcalypso.localhost%3A3000%2Fjetpack%2Fconnect%2Fauthorize%3Fclient_id%3D212713831%26redirect_uri%3Dhttps%253A%252F%252Ffalse-shrike.jurassic.ninja%252Fwp-admin%252Fadmin.php%253Fhandler%253Djetpack-connection-webhooks%2526action%253Dauthorize%2526_wpnonce%253D0e6a75dad7%2526redirect%253Dhttps%25253A%25252F%25252Ffalse-shrike.jurassic.ninja%25252Fwp-admin%25252Fadmin.php%25253Fpage%25253Djetpack%26state%3D1%26scope%3Dadministrator%253A3a6e1fb4a0d3fca03c4b16915582698f%26user_email%3Ddemo%2540example.org%26user_login%3Ddemo%26jp_version%3D11.5.1%26secret%3DvR0f4dhOxJWjtiika7pL3ns8z8JOsPHS%26blogname%3DFalse%2BShrike%26site_url%3Dhttps%253A%252F%252Ffalse-shrike.jurassic.ninja%26home_url%3Dhttps%253A%252F%252Ffalse-shrike.jurassic.ninja%26site_icon%26site_lang%3Den_US%26site_created%3D2022-11-23%2B12%253A55%253A29%26allow_site_connection%3D1%26locale%3Den%26_ui%3D335273%26_ut%3Dwpcom%253Auser_id%26_as%3Dwp-cli%26purchase_nonce%3DslDQZ1DY%26_wp_nonce%3D80f05dcb20%26redirect_after_auth%3Dhttps%253A%252F%252Ffalse-shrike.jurassic.ninja%252Fwp-admin%252Fadmin.php%253Fpage%253Djetpack%26site%3Dhttps%253A%252F%252Ffalse-shrike.jurassic.ninja&email_address=demo@example.org&from=%5Bunknown%5D&allow_site_connection=1`

### CODE REVIEW

File: client/layout/logged-out.jsx [hunk 1]
```diff
@@ -280,7 +280,7 @@ export default withCurrentRoute(
 		const isPopup = '1' === currentQuery?.is_popup;
 		const noMasterbarForSection =
 			! isWooOAuth2Client( oauth2Client ) &&
-			[ 'accept-invite', 'signup', 'jetpack-connect' ].includes( sectionName );
```
Comment (alshakero) [inline client/layout/logged-out.jsx:283]: Fishy, just making sure it's intentional.

File: client/my-sites/invites/invite-accept-logged-out/index.jsx [hunk 2]
```diff
@@ -156,28 +154,6 @@ class InviteAcceptLoggedOut extends Component {
...
-		return (
-			<div className="invite-logged-out-header">
-				<WordPressLogo size={ 24 } />
-
-				<Button href={ '/log-in?redirect_to=' + window.location.href } variant="link">
-					<span>{ this.props.translate( 'Log in' ) }</span>
-				</Button>
-			</div>
-		);
-	};
```
Comment (alshakero) [inline client/my-sites/invites/invite-accept-logged-out/index.jsx:179]: Would appreciate an explanation of this change. Can't figure it out.

File: client/signup/config/flows-pure.js [hunk 3]
```diff
@@ -30,124 +40,135 @@ export function generateFlows( {
 		},
 		{
 			name: 'account',
-			steps: [ 'user' ],
+			steps: [ userSocialStep ],
```
Comment (alshakero) [inline client/signup/config/flows-pure.js:43]: ❤️

File: client/signup/steps/user/index.jsx [hunk 4]
```diff
@@ -259,7 +260,23 @@ export class UserStep extends Component {
...
+					} );
+				} else {
+					subHeaderText = translate(
+						'First, create your WordPress.com account. Have an account? {{a}}Log in{{/a}}',
+						{
+							components: { a: <a href={ loginUrl } rel="noopener noreferrer" /> },
+						}
+					);
+				}
+			}
```
Comment (alshakero) [inline client/signup/steps/user/index.jsx:279]: Would appreciate a line about this change.

File: client/signup/steps/user/index.jsx [hunk 5]
```diff
@@ -450,7 +467,7 @@ export class UserStep extends Component {
 			return translate( 'Let’s get you signed up.' );
 		}
 
-		if ( ! headerText ) {
+		if ( isSocialFirst && ! headerText ) {
```
Comment (alshakero) [inline client/signup/steps/user/index.jsx]: Is there a reason `isSocialFirst` shouldn't have a header fallback?

File: client/signup/steps/user/index.jsx [hunk 6]
```diff
@@ -594,6 +617,26 @@ export class UserStep extends Component {
...
+				</Button>
+			);
+		}
+	}
+
+	getIsSticky() {
+		if ( this.props.isSocialFirst ) {
+			return false;
+		}
+	}
```
Comment (alshakero) [inline client/signup/steps/user/index.jsx:638]: This function always returns `falsy`. Is that intentional?

**Review by alshakero — COMMENTED** [review]
Beautiful. I like this approach much more!! Left some questions.

File: client/layout/logged-out.jsx (hunk 1 above)
Comment (escapemanuele) [reply client/layout/logged-out.jsx:283]: Yes, I left the `accept-invite` part out of the restyling, to try to limit the complexity of all of this. Also the approach we used involved a lot of CSS that I was not sure I was able to manage well with all the other changes now.
After this PR we will update that part too 👍

File: client/my-sites/invites/invite-accept-logged-out/index.jsx (hunk 2 above)
Comment (escapemanuele) [reply client/my-sites/invites/invite-accept-logged-out/index.jsx:179]: Same as before, these changes were added in the first Social signup PR, and I'm keeping the accept-invite out of the new social login as I want to work on that with a bit more accuracy.

File: client/signup/steps/user/index.jsx (hunk 4 above)
Comment (escapemanuele) [reply client/signup/steps/user/index.jsx:279]: Code starting from `const { queryObject } = this.props;` is what was already there in the User step. 
`subHeaderText = '';` was added when making this a Social First signup. Now I'm reintroducing the old subheader when this is the old User step

File: client/signup/steps/user/index.jsx (hunk 6 above)
Comment (escapemanuele) [reply client/signup/steps/user/index.jsx:638]: Yes! Without this, [this line](https://github.com/Automattic/wp-calypso/blob/trunk/client/signup/step-wrapper/index.jsx#L206) would set the `sticky` prop on `ActionButtons` to null,  positioning the Log In custom button to the bottom on mobile:

image

File: client/signup/steps/user/index.jsx (hunk 5 above)
Comment (escapemanuele) [reply client/signup/steps/user/index.jsx]: Updated as I think it should always have the same header when social first

File: client/signup/steps/user/index.jsx (hunk 6 above)
Comment (alshakero) [reply client/signup/steps/user/index.jsx:638]: Can we make it explicit and return `false` in other cases?

**Comment by renancarvalho** [comment]
Hi Emanuele, I got this using live link, they look a bit different from what you posted. 

Gravatar:
image

p2:
image

**Comment by escapemanuele** [comment]
Thank you for testing @renancarvalho, in the first you have to click `Create one` and in the second one `Create a new account`, to see the same pages!

**Comment by heavyweight** [comment]
Some notes from my testing

**Testing the onboarding-pm flow**
✅ onboarding-pm

**Flows from stepper**
✅ /setup/newsletter
✅ /setup/link-in-bio and 
✅ /setup/domain-transfer
I tested with the ff turned off based on the url using `?flags=-signup/social-first`
✅ /setup/newsletter
✅ /setup/link-in-bio and 
⛔️ /setup/domain-transfer. I'm unable to test this with the ff off based on the URL. Can someone check it?

**Checks for URLs that should not be affected**
✅ Gravatar
✅ P2
✅ Crowdsignal
✅ Akismet
✅ Jetpack connect

**Check existing flows**
Went through these flows with the ff on or off when applicable
✅ hosting
✅ account
✅ business
✅ premium
✅ personal
✅ pro
✅ starter
✅ free
⛔️ with-theme. I was unable to test it due to dependencies. Can someone else try?
✅ with-theme-assembler
⛔️ with-plugin. I was unable to test it due to dependencies. Can someone else try?
✅ onboarding
✅ onboarding-2023-pricing-grid
⛔️ domain-transfer. I am unable to test this with the ff off based on the URL. Can someone else try?
✅ onboarding-media
✅ import
✅ with-add-ons
✅ onboarding-with-email
✅ onboarding-registrationless
✅ desktop
✅ pressable-nux
✅ rewind-switch
⛔️ rewind-setup. I was unable to test it. Can someone else try?
✅ rewind-auto-config. I was unable to test it due to dependencies. Can someone else try?
✅ clone-site. I was unable to test it due to dependencies. Can someone else try?
✅ ecommerce
✅ commerce-monthly
✅ wpcc
⛔️ p2v1. Getting redirect loop here with ff on or off
✅ videopress-account
✅ p2
✅ domain
⛔️ site-selected. I was unable to test it due to dependencies. Can someone else try?
⛔️ launch-site. I was unable to test it due to dependencies. Can someone else try?
✅ reader
✅ crowdsignal
✅ plan-no-domain
⛔️ launch-only. I was unable to test it due to dependencies. Can someone else try?
✅ business-monthly
✅ premium-monthly
✅ personal-monthly
✅ setup-site
✅ do-it-for-me
✅ do-it-for-me-store
⛔️ website-design-services. I was unable to test it due to dependencies. Can someone else try?
⛔️ site-content-collection. I was unable to test it due to dependencies. Can someone else try?
⛔️ woocommerce-install. I was unable to test it due to dependencies. Can someone else try?
✅ ecommerce-2y
✅ business-2y
✅ premium-2y
✅ personal-2y

**Comment by renancarvalho** [comment]
**I Will keep updating this comment as I test**


Checking what @heavyweight could not check.

**Flows from stepper**
✅ Domain transfer using [this url](https://container-kind-sanderson.calypso.live/setup/domain-transfer/intro)
⛔️ rewind-setup seems broken, but also in production, no errors on console.

**Comment by renancarvalho** [comment]
**I Will keep updating this comment as I test**


Checking what @heavyweight could not check.

**Flows from stepper**
✅ Domain transfer using [this url](https://container-kind-sanderson.calypso.live/setup/domain-transfer/intro)

**Comment by escapemanuele** [comment]
Thanks for testing team!

@heavyweight
> I was unable to test it due to dependencies

If that happens, you can check the console for something like
```
Error: with-theme did not provide the query dependencies [theme] it is configured to.
```

And, in this case, append the needed parameter => `LiveLink/start/with-theme?theme=random-theme`

I'll better check `/setup/domain-transfer`, `rewind-setup` and the other that appear to be broken!

**Comment by escapemanuele** [comment]
Also, @heavyweight & @renancarvalho, I think that `rewind-setup` is accessible only when logged in. Moreover, it has not an user step, so it shouldn't be affected here 🤔

**Comment by agrullon95** [comment]
## Testing notes (still updating): 

Edge cases (Gravatar, P2, Crowdsignal, Akismet, Jetpack Connect) are all displaying their own signup screen.

Tested all the flows from [/config/flows-pure.js](https://github.com/Automattic/wp-calypso/blob/update/componentize-signup-social-first/client/signup/config/flows-pure.js) and the ones that include the `userSocialStep` step are correctly navigating to the new social login screen. 

Some flows didn't load correctly, but expected behavior since I was missing some of the query param dependencies according to the console error.
For example, `with-plugin` didn't load correctly and this was the console error:

```
with-plugin:
Error: with-plugin did not provide the query dependencies [plugin,billing_period] it is configured to.
    at Se._assertFlowProvidedDependenciesFromConfig (signup.499b51c6537d6cd2959c.min.js:1:21886)
   ```
  
I dont think this is related but I ran into a blank screen issue with `rewind-setup` (EDIT: Looks like you already addressed that we need to be logged in) and infinite page loading loop with `p2v1`

**Review by alshakero — APPROVED** [review]
Did a final smoke test and tested 5 random flows and all works as expected.

**Pull request merged**